src/notificationmanager.cpp

```diff
@@ -771,3 +768,67 @@ void NotificationManager::slotNgfEventFinished(quint32 id)
         m_ngfEvent = 0;
 }
 
+void NotificationManager::slotVoicemailWaitingChanged(bool waiting)
+{
+    uint currentId = 0;
+
+    // See if there is a current notification for voicemail waiting
+    QList<QObject*> notifications = Notification::notifications();
+    foreach (QObject *o, notifications) {
+        Notification *n = static_cast<Notification*>(o);
+        if (n->category() == voicemailWaitingCategory) {
+            if (waiting) {
+                // The notification is already present; do nothing
+                currentId = n->replacesId();
+            } else {
+                // Close this notification
+                n->close();
+            }
+        }
+    }
+    qDeleteAll(notifications);
```

Comment:
Maybe add `notifications.clear();` below the `qDeleteAll()` call to make sure nobody accidentally touches any dangling pointers in the future.